tests/test_auth.py

```diff
@@ -0,0 +1,186 @@
+    "algorithm": "sha256"
+}
+
+INVALID_USER = {
+    "id": "babadie@mozilla.com",
+    "key": "INVALID_SECRET",
+    "algorithm": "sha256"
+}
+
+
+def mock_hawk(method, url='/', body='', user=VALID_USER):
+    '''
+    Helper to create an HAWK header towards mock server
+    '''
+    full_url = 'http://localhost' + url
+    sender = Sender(user, full_url, method, body, content_type='application/json')
+    return {
+        'Authorization': sender.request_header,
+        'Content-Type': 'application/json',
+    }
+
+
+class TestAuthentication(unittest.TestCase):
+    '''
+    Test the authentication process through Hawk
+    '''
+    def setUp(self):
+        logging.basicConfig()
+
+        # Setup test settings
+        frontline.settings = {
+            'whitelist': ['test-index', 'protected-data', ],
+            'elasticsearch': [{
+                "host":"http://unittest",
+                "port":9200
+            }],
+        }
+        frontline.auth.load_users([{
+          "hawk": VALID_USER,
+          "resources": ['test-index', ],
+        }])
+
+        # Setup flask app Client
+        frontline.app.testing = True
+        self.client = frontline.app.test_client()
+
+        # Setup responses
+        mock_es = re.compile('http://unittest:9200/.*')
+        responses.add(
+            responses.GET,
+            mock_es,
+            body='{}',
+            content_type='application/json',
+        )
+        responses.add(
+            responses.HEAD,
+            mock_es,
+            body='',
+            content_type='application/json',
+        )
+
+    @responses.activate
+    def test_head(self):
+        '''
+        HEAD request should always work when authenticated
+        '''
+
+        # No auth
+        r = self.client.head('/')
+        self.assertEqual(r.status_code, 403)
+
+        # Invalid auth
+        r = self.client.head('/', headers={'Authorization': 'Invalid token'})
+        self.assertEqual(r.status_code, 403)
+
+        # Valid auth
+        valid_hawk = mock_hawk('HEAD')
+        r = self.client.head('/', headers=valid_hawk)
+        self.assertEqual(r.status_code, 200)
+
+        # Replay not allowed
+        r = self.client.head('/', headers=valid_hawk)
+        self.assertEqual(r.status_code, 403)
+
+        # Invalid secret
+        invalid_hawk = mock_hawk('HEAD', user=INVALID_USER)
+        r = self.client.head('/', headers=invalid_hawk)
+        self.assertEqual(r.status_code, 403)
+
+    @responses.activate
+    def test_get(self):
+        '''
+        Test some GET requests with authentication
+        '''
+        search_url = '/test-index/_search'
+
+        # No auth, no query
+        r = self.client.get(search_url )
+        self.assertEqual(r.status_code, 403)
+
+        # Valid auth but missing query
+        valid_hawk = mock_hawk('GET', search_url, '{}')
+        r = self.client.get(search_url, data='{}', headers=valid_hawk)
+        self.assertEqual(r.status_code, 400)
```

Comment:
I am having trouble understanding this test: The "request" goes to the `self.client` directly, but the `app.catch_all()` will call out to `http://unittest:9200/test-index/_search`, and I do not know where that goes!

Reply:
OK, I think I got it.  The `responses` lib is interesting! ...and a bit scary :)